cou/utils/openstack.py

```diff
@@ -13,118 +13,129 @@
+    For charm gnocchi it was used cmadison.
+
+    [0] https://releases.openstack.org/
+    [1] https://docs.openstack.org/charm-guide/latest/project/charm-delivery.html
+
+    :return: Ordered dictionary containing the version and the possible OpenStack release.
+    :rtype: OrderedDict
+    """
+    with open(CSV_FILE) as csv_file:
+        openstack_lookup = OrderedDict()
+        csv_reader = csv.reader(csv_file, delimiter=",")
+        header = next(csv_reader)
+        for row in csv_reader:
+            service_dict: defaultdict[str, Any] = defaultdict(OrderedDict)
+            service = row[SERVICE_COLUMN_INDEX]
+            for column_index in range(VERSION_START_COLUMN_INDEX, len(row), 2):
+                os_version, _ = header[column_index].split("-")
+                min_version = row[column_index]
+                max_version = row[column_index + 1]
+                service_dict[os_version] = VersionRange(
+                    Version(min_version),
+                    Version(max_version),
+                )
+            openstack_lookup[service] = service_dict
+    for charm_type, charms in CHARM_TYPES.items():
+        for charm in charms:
+            openstack_lookup[charm] = openstack_lookup[charm_type]
+    return openstack_lookup
+
+
+class OpenStackLookup(object):
+    _OPENSTACK_LOOKUP: OrderedDict = OrderedDict()
+
+    def __new__(cls):
+        """Magic method for create the OpenStackLookup object.
+
+        :return: OpenStackLookup object
+        :rtype: OpenStackLookup
+        """
+        if not cls._OPENSTACK_LOOKUP:
+            cls._OPENSTACK_LOOKUP = generate_openstack_lookup()
+        return super().__new__(cls)
+
+    def get_compatible_openstack_codenames(
+        self, component: str, version: str
+    ) -> List[Optional[str]]:
+        """Get the possible OpenStack codenames based on the component and version.
+
+        :param component: Name of the component. E.g: "keystone"
+        :type component: str
+        :param version: Version of the component. E.g: "17.0.2"
+        :type version: str
+        :return: Return a list with the compatible OpenStack codenames.
+        :rtype: List[Optional[str]]
+        """
+        possible_os_releases: List[Optional[str]] = []
+        if not self._OPENSTACK_LOOKUP.get(component):
+            logging.warning(
+                (
+                    "Not possible to find a compatible OpenStack codename for "
+                    "charm: %s with workload_version: %s"
+                ),
+                component,
+                version,
+            )
+            return possible_os_releases
+        for openstack_release, version_range in self._OPENSTACK_LOOKUP[component].items():
+            if version in version_range:
+                possible_os_releases.append(openstack_release)
+        return possible_os_releases
+
+
+openstack_lookup = OpenStackLookup()
```

Comment:
Running code which has side effects while importing module is bad practice. For example if they do not have csv file it will throw file not found error when they import module.  Maybe something like:
```python
class OpenStackCodenameLookup(object):
    _OPENSTACK_LOOKUP: OrderedDict = OrderedDict()
    _DEFAULT_CSV_FILE = "cou/utils/openstack_lookup.csv"
   
    @classmethod
    def initialize_lookup(cls, resource: Optional[File]):
        file_to_import = resource
        if not file_to_import:
            file_to_import = _DEFAULT_CSV_FILE
        cls._OPENSTACK_LOOKUP = _generate_lookup(file_to_import)

    @classmethod
    def lookup(cls, component: str, version: str) -> List[Optional[str]]:
       if not initialized:
           initialize_lookup
        ...
    @classmethod
    def _generate_lookup(cls, file_to_import):
        ...
```

Reply:
yes, I agree that an import should not yield an exception. However let's consider that the csv file would be part of the codebase, so failing to find it and read it is equivalent to failing to find or read any other Python file we're delivering as part of this snap